metricsmap: fix Prometheus exporter

jibi opened the pull request:
Fix the Prometheus exporter as it was not aggregating correctly the
multiple entries that make up a value.

While at it, simplify the logic used to iterate the map by replacing it
with a call to `Metrics.DumpWithCallback()`.

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

pchaigno commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -212,69 +206,48 @@ func updateMetric(getCounter func() (prometheus.Counter, error), newValue float6
-		// CPU is zero, it never iterates over the next non-zero value.
-		for i := 0; i < possibleCpus; i++ {
-			// Increment Prometheus metrics here.
-			updatePrometheusMetrics(&nextKey, &entry[i])
```
Since you added a new comment to that regard, I thought that the bug was that we simply forgot to iterate on per-CPU values, but it seems we were already doing that here :point_up: 
Am I missing something?

jibi replied:
This is slightly subtle as I'm changing the logic from:

```go
func updatePrometheusMetrics(value) {
  updateMetric(value.Packets)
  updateMetric(value.Bytes)
}

func SyncMetrics() {
  for _, value := range values {
    updatePrometheusMetrics(value)
  }
}
```

to:

```go
func updatePrometheusMetrics(values) {
  for _, value := range values {
    packets += value.Packets
    bytes   += value.Bytes
  }

  updateMetrics(packets)
  updateMetrics(bytes)
}


func SyncMetrics() {
  updatePrometheusMetrics(values)
}
```

What `updateMetric` does is:
https://github.com/cilium/cilium/blob/8bf3ed83cfbbbf8c9da2d7c57c1b8debecc6b35d/pkg/maps/metricsmap/metricsmap.go#L199-L210

so:

* diff the `packets`/`bytes` value it receive with the one stored in Prometheus
* increment the Prometheus counter by the delta it just calculated

Since the logic in `updateMetric` is not linear, it will work correctly only if the `packets`/`bytes` counter it receives is the _sum_ of all the per-CPU values for a given key (rather than what we are doing now, which is calling `updateMetric()` on every single per-CPU value for a given key)

pchaigno replied:
Ah, got it. Thanks for the explanation!

It would probably make sense to explain that in the commit message, or to break it down in two commits to separate the fix from the refactoring.

jibi replied:
Good shout, let me split the commit

jibi committed:
metricsmap: simplify logic to iterate map in SyncMetricsMap

Simplify the logic used to iterate the metrics map by replacing it with
a call to Metrics.DumpWithCallback().

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

jibi committed:
metricsmap: fix Prometheus exporter

Fix the Prometheus exporter as it was not aggregating correctly the
multiple entries that make up a value.

This commit changes the logic from:

    func updatePrometheusMetrics(value) {
      updateMetric(value.Packets)
      updateMetric(value.Bytes)
    }

    func SyncMetrics() {
      for _, value := range values {
	updatePrometheusMetrics(value)
      }
    }

to:

    func updatePrometheusMetrics(values) {
      for _, value := range values {
	packets += value.Packets
	bytes   += value.Bytes
      }

      updateMetrics(packets)
      updateMetrics(bytes)
    }

    func SyncMetrics() {
      updatePrometheusMetrics(values)
    }

What updateMetric does is:

* diff the packets/bytes value it receive with the one stored in
  Prometheus
* increment the Prometheus counter by the delta it just calculated

Since the logic in updateMetric is not linear, it will work correctly
only if the packets/bytes counter it receives is the sum of all the
per-CPU values for a given key (rather than what we are doing now, which
is calling updateMetric() on every single per-CPU value for a given key)

Fixes: a59105352 ("pkg/maps/metricsmap: userspace aggregation of BPF_PER_CPU_HASH_MAP metrics map.")

Signed-off-by: Gilberto Bertin gilberto@isovalent.com
